sklearn/_build_utils/__init__.py

```diff
@@ -33,3 +35,42 @@ def atlas_not_found(blas_info_):
         cblas_libs = blas_info.pop('libraries', [])
 
     return cblas_libs, blas_info
+
+
+def get_cython_source(filename):
+    is_c_filename = filename.endswith('.c')
+    is_cpp_filename = filename.endswith('.cpp')
+
+    # files in src are .c and .cpp files that are not cython-generated
+    if ('src' + os.sep) in filename and (is_c_filename or is_cpp_filename):
```

Comment:
How about checking if there exists a ``pyx`` file with the same name in the same folder?